File: Asgard_Test/tests_Heimdall/benchmarks/corpus/taint/tp_sa4_helper_call_site_sensitivity.py
"""TP: SA4 context-sensitivity -- the SAME helper is called from a clean
call site (a literal) and a sinking call site (tainted param). These must
NOT be conflated: exactly one flow, from the tainted call site."""

import subprocess


def run_cmd(host):
    subprocess.run("ping -c 1 " + host, shell=True)


def clean_call():
    run_cmd("localhost")


def tainted_call():
    host = request.args.get("host")
    run_cmd(host)

File: Asgard_Test/tests_Heimdall/benchmarks/corpus/taint_js/tp_sa4_helper_call_site_sensitivity.js
function lookupUser(id) {
    db.query('SELECT * FROM users WHERE id=' + id);
}

// SA4 context-sensitivity: the SAME helper is called from a clean call
// site (a literal) and a sinking call site (tainted param). These must
// NOT be conflated -- exactly one flow (from the tainted call site).
app.get('/safe', (req, res) => {
    lookupUser('admin');
    res.send('ok');
});

app.get('/user', (req, res) => {
    const id = req.query.id;
    lookupUser(id);
    res.send('ok');
});
